components/hab/src/command/plan/render.rs

```diff
@@ -0,0 +1,176 @@
+        ui.begin(format!("Importing default.toml: {}", &default_toml_path))?;
+    }
+
+    // we should always have a default.toml, would be nice to "autodiscover" based on package name,
+    // for now assume we're working in the plan dir if --default-toml not passed
+    let default_toml = read_to_string(&default_toml_path)
+        .expect(&format!("Something went wrong reading: {}", &default_toml_path));
+    // merge default into data struct
+    merge(&mut data, toml_to_json(&default_toml));
+
+    // import default.toml values, convert to JSON
+    // ui.begin(format!("Importing user.toml: {}", &user_toml_path));
+    let user_toml = match user_toml_path {
+        Some(path) => {
+            if !(quiet) {
+              // print helper message, maybe only print if '--verbose'? how?
+              ui.begin(format!("Importing user.toml: {}", path.to_string()))?;
+            }
+            read_to_string(path.to_string())
+                .expect(&format!("Something went wrong reading: {}", path.to_string()))
+        },
+        None => "".to_string(),
+    };
+    // merge default into data struct
+    merge(&mut data, toml_to_json(&user_toml));
+
+    // read mock data if provided
+    let mock_data = match mock_data_path {
+        Some(path) => {
+            if !(quiet) {
+                // print helper message, maybe only print if '--verbose'? how?
+                ui.begin(format!("Importing override file: {}", path.to_string()))?;
+            }
+            read_to_string(path.to_string())
+              .expect(&format!("Something went wrong reading: {}", path.to_string()))
+        },
+        // return an empty json block if '--mock-data' isn't defined.
+        // this allows us to merge an empty JSON block
+        None => "{}".to_string(),
+    };
+    // merge mock data into data
+    merge(&mut data, serde_json::from_str(&mock_data).unwrap());
+
+    // create a template renderer
+    let mut renderer = TemplateRenderer::new();
+    // register our template 
+    renderer
+        .register_template_string(&template_path, &template)
+        .expect("Could not register template content");
+    // render our JSON override in our template.
+    let rendered_template = renderer.render(&template_path, &data).ok().unwrap();
+    
+    if print {
+        if !(quiet) {
+          ui.br()?;
+          ui.warn(format!("###======== Rendered template: {}", &template_path))?;
+        }
+
+        println!("{}", rendered_template);
+
+        if !(quiet) {
+          ui.warn(format!("========### End rendered template: {}", &template_path))?;
+        }
+    }
+
+    // if not no render dir (aka "unless no_render_dir == true")
+    if !(no_render_dir) {
+      // Render our template file
+      create_with_template(ui, &format!("{}/{}", render_dir, file_name), &rendered_template, quiet)?;
+    }
+
+    if !(quiet) {
+      ui.br()?;
+    }
+    // not really sure this is correct...
```

Comment:
👍

Reply:
well it was actually the next line returning 

```
Ok(())
```

What is that :point_up:.  I undertsand that `Ok` is a `std::result` type, and indicates success.  but why do we pass `Ok` the `()`?

Reply:
`()` is [the unit type](http://doc.rust-lang.org/1.32.0/std/primitive.unit.html), similar to `void` in some other languages. In this case it's necessary because the return type of `start` is `Result<()>`, which means "this function may return an error, but if not, there's no additional information to return".

Reply:
ahhh, ok!  Thanks!